crates/cubecl-cpp/src/shared/instruction.rs

```diff
@@ -651,8 +653,334 @@ for ({i_ty} {i} = {start}; {i} {cmp} {end}; {increment}) {{
+            "__nv_cvt_{in_ty}_to_{out_ty}({in_value}, __NV_NOSAT, __NV_{interpretation})",
+        )
+    })
+}
+
+fn float_to_packed<D: Dialect>(input: Variable<D>, i: usize, packing: usize) -> String {
+    match input.elem() {
+        Elem::TF32 | Elem::F32 => {
+            let i = i * packing;
+            if packing > 1 {
+                format!("float2 {{ {}, {} }}", input.index(i), input.index(i + 1))
+            } else {
+                format!("{}", input.index(i))
+            }
+        }
+        Elem::F64 => {
+            let i = i * packing;
+            if packing > 1 {
+                format!("double2 {{ {}, {} }}", input.index(i), input.index(i + 1))
+            } else {
+                format!("{}", input.index(i))
+            }
+        }
+        Elem::F16 | Elem::F16x2 | Elem::BF16 | Elem::BF16x2 => format!("{}", input.index(i)),
+        _ => unreachable!(),
+    }
+}
+
+fn cast_minifloat_to_half<D: Dialect>(
+    f: &mut fmt::Formatter,
+    input: Variable<D>,
+    out: Variable<D>,
+) -> fmt::Result {
+    let in_opt = input.optimized();
+    let out_opt = out.optimized().item();
+
+    let (in_ty, interpretation) = match in_opt.elem() {
+        Elem::FP4(kind) => ("fp4", format!("{kind:?}")),
+        Elem::FP4x2(kind) => ("fp4x2", format!("{kind:?}")),
+        Elem::FP6(kind) => ("fp6", format!("{kind:?}")),
+        Elem::FP6x2(kind) => ("fp6x2", format!("{kind:?}")),
+        Elem::FP8(kind) => ("fp8", format!("{kind:?}")),
+        Elem::FP8x2(kind) => ("fp8x2", format!("{kind:?}")),
+        _ => unreachable!("can only cast minifloat"),
+    };
+
+    let out_ty = match out_opt.elem() {
+        Elem::F16 => "halfraw",
+        Elem::F16x2 => "halfraw2",
+        _ => unreachable!("out type must be half"),
+    };
+
+    handle_unroll(f, out, |f, i| {
+        let input = in_opt.index(i);
+        write!(
+            f,
+            "{}(__nv_cvt_{in_ty}_to_{out_ty}({input}, __NV_{interpretation}))",
+            out_opt.elem()
+        )
+    })
+}
+
+fn cast_scale_to_bfloat<D: Dialect>(
+    f: &mut fmt::Formatter,
+    input: Variable<D>,
+    out: Variable<D>,
+) -> fmt::Result {
+    let in_opt = input.optimized();
+    let out_opt = out.optimized().item();
+
+    let in_ty = match in_opt.elem() {
+        Elem::FP8(_) => "e8m0",
+        Elem::FP8x2(_) => "e8m0x2",
+        _ => unreachable!("must be scaling factor in e8m0 format"),
+    };
+
+    let out_ty = match out_opt.elem() {
+        Elem::BF16 => "bf16raw",
+        Elem::BF16x2 => "bf162raw",
+        _ => unreachable!("out type must be half"),
+    };
+
+    handle_unroll(f, out, |f, i| {
+        let input = in_opt.index(i);
+        write!(
+            f,
+            "{}(__nv_cvt_{in_ty}_to_{out_ty}({input}))",
+            out_opt.elem()
+        )
+    })
+}
+
+fn handle_unroll<D: Dialect>(
+    f: &mut fmt::Formatter,
+    out: Variable<D>,
+    mut op: impl FnMut(&mut fmt::Formatter, usize) -> fmt::Result,
+) -> fmt::Result {
```

Comment:
All of those cast functions could be moved into another module just for that.

Reply:
Agreed, it really balooned to a much larger amount of code than expected once I realized how annoying the CUDA APIs are